controllers/nstemplateset/space_roles.go

```diff
@@ -0,0 +1,136 @@
+	if err != nil {
+		return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusProvisionFailed, err,
+			"failed to list namespaces for workspace '%s'", nsTmplSet.Name)
+	}
+	logger.Info("ensuring space roles", "namespace_count", len(nss), "role_count", len(nsTmplSet.Spec.SpaceRoles))
+	changed := false
+	for _, ns := range nss {
+		// space roles to apply now
+		lastAppliedSpaceRoleObjs, err := r.getLastAppliedSpaceRoles(&ns)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdateForSpaceRolesFailure(logger, nsTmplSet, err, "failed to retrieve last applied space roles")
+		}
+		spaceRoleObjs, err := r.getSpaceRoles(&ns, nsTmplSet.Spec.SpaceRoles)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdateForSpaceRolesFailure(logger, nsTmplSet, err, "failed to retrieve space roles to apply")
+		}
+
+		c, err := deleteOutdatedObjects(logger, r.Client, lastAppliedSpaceRoleObjs, spaceRoleObjs)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusUpdateFailed, err, "failed to delete redundant objects in namespace '%s'", ns.Name)
+		}
+		changed = changed || c
+		// labels to apply on all new objects
+		var labels = map[string]string{
+			toolchainv1alpha1.ProviderLabelKey: toolchainv1alpha1.ProviderLabelValue,
+			toolchainv1alpha1.OwnerLabelKey:    nsTmplSet.GetName(), // TODO: is it needed?
+		}
+		logger.Info("creating space role objects")
+		for _, obj := range spaceRoleObjs {
+			logger.Info(fmt.Sprintf("creating/updating %s/%s", obj.GetNamespace(), obj.GetName()))
+		}
+		// create (or update existing) objects based the tier template
+		c, err = applycl.NewApplyClient(r.Client, r.Scheme).Apply(spaceRoleObjs, labels)
+		if err != nil {
+			return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusNamespaceProvisionFailed, err, "failed to provision namespace '%s' with space roles", ns.Name)
+		}
+		changed = changed || c
+		if changed {
+			// store the space roles in an annotation at the namespace level, so we know what was applied and how to deal with
+			// diffs when the space roles are changed (users added or removed, etc.)
+			sr, _ := json.Marshal(nsTmplSet.Spec.SpaceRoles)
+			if ns.Annotations == nil {
+				ns.Annotations = map[string]string{}
+			}
+			ns.Annotations[toolchainv1alpha1.LastAppliedSpaceRolesAnnotationKey] = string(sr)
+			if err := r.Client.Update(context.TODO(), &ns); err != nil {
+				return false, r.wrapErrorWithStatusUpdate(logger, nsTmplSet, r.setStatusProvisionFailed, err,
+					"failed update namespace annotation")
+			}
+		}
+	}
+	return changed, nil
+}
+
+// Get the space role objects from the templates specified in the given `spaceRoles`
+// Returns the objects indexed by kind and name, or an error if something wrong happened when processing the templates
+func (r *spaceRolesManager) getSpaceRoles(ns *corev1.Namespace, spaceRoles []toolchainv1alpha1.NSTemplateSetSpaceRole) ([]runtimeclient.Object, error) {
```

Comment:
I would maybe name it `getSpaceRolesObjects` but I don't insist on it. 

btw, what do you mean by: 
```
// Returns the objects indexed by kind and name
```

Reply:
> I would maybe name it getSpaceRolesObjects but I don't insist on it.

sure, let me do that

> btw, what do you mean by:

ah never mind, this comment is outdated, I'll remove it

Reply:
done in a58732d